core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/DetermineJoinDistributionType.java

```diff
@@ -161,7 +219,21 @@ private JoinNode getSizeBasedJoin(JoinNode joinNode, Context context)
 
         if (isLeftSideSmall) {
             // left side is small enough, but must be partitioned
-            return joinNode.flipChildren().withDistributionType(PARTITIONED);
+            return flippedJoin.withDistributionType(PARTITIONED);
+        }
+
+        // Flip join sides if left side is smaller than the right by more than SIZE_DIFFERENCE_THRESHOLD times.
+        // We use 8x factor because the reduction in size from aggregation and filtering
+        // may not be fully accounted here.
+        // We use getApproximateOutputSizeInBytes instead of getSourceTablesSizeInBytes to account for the effects of
+        // the operators between join and table scan as much as possible when comparing the size of the join sides.
+        double leftSourceSize = getApproximateOutputSizeInBytes(joinNode.getLeft(), context);
+        double rightSourceSize = getApproximateOutputSizeInBytes(joinNode.getRight(), context);
+        if (leftSourceSize * SIZE_DIFFERENCE_THRESHOLD < rightSourceSize) {
+            if (mustReplicate(flippedJoin, context)) {
+                return flippedJoin.withDistributionType(REPLICATED);
+            }
+            return flippedJoin.withDistributionType(PARTITIONED);
         }
```

Comment:
> I've simplified this is a bit more to remove the non-flipped case as that's not really needed.

Why? How do you know `getSyntacticOrderJoin` wont flip join sides? It would be better to make logic here explicit

Reply:
That's based on reading of the current code, it does not flip join sides. However, I've restored the explicit branch now.